### PR TITLE
Provide initial end-to-end tests

### PR DESCRIPTION
**Pull request by wphicks**
Provide a set of end-to-end tests to test full range of FIL backend functionality
Provide a wrapper script for running tests locally
Provide easier debug builds through Dockerfile build args and environment variables

Initial PR for resolution of #26

### CODE REVIEW

**Comment by wphicks** [comment]
The goal of this PR is to provide an initial, functional set of tests to build upon with later CI/testing work. Later PRs can cover more granular testing and hooking up Gitlab CI as laid out in #26.

**Comment by wphicks** [comment]
Since I don't think anyone has started to look at this, I went ahead and pushed the reorganization necessary to make this work with CI. I'll go ahead and trigger the pipeline for it so that we can see that we're currently passing with these changes.

**Comment by hcho3** [comment]
@wphicks When do you want me to look at this? CI should of course verify that the pipeline runs, but I can comment on organization and design.

**Comment by wphicks** [comment]
@hcho3 If you can start taking a look at it right away, that would be great! Just added that last batch of commits to hopefully streamline the review process for getting the whole CI system up and running

File: qa/L0_e2e/test_model.py [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+        retries=3,
+        host='localhost',
+        http_port=STANDARD_PORTS['http'],
+        grpc_port=STANDARD_PORTS['grpc']):
+
+    if batch_sizes is None:
+        batch_sizes = []
+        if total_rows > 128:
+            batch_sizes.append(128)
+        batch_sizes.append(1)
```
Comment (hcho3) [inline qa/L0_e2e/test_model.py:222]:
```suggestion
        batch_sizes = [1]
        if total_rows > 128:
            batch_sizes.append(128)
```

File: qa/L0_e2e/test_model.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+
+    throughput = len(batch_sizes) * total_rows / total
+    per_sample = 1 / throughput
+
+    request_latency = 0
+    row_latency = 0
+    for indices, (triton_result, batch_latency) in results:
+        if triton_result is None:
+            raise RuntimeError(
+                f'Prediction failed with error:\n\n{batch_latency}'
```
Comment (hcho3) [inline qa/L0_e2e/test_model.py:425]: Maybe add a comment indicating that, when an error occurs, `batch_latency` contains the error code?

**Review by hcho3 — COMMENTED** [review]
The overall testing suite looks good to me. Some comments:

File: qa/test_entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+models=()
+
+echo 'Generating example models...'
+models+=( $(python /L0_e2e/generate_example_model.py \
+  --name xgboost \
+  --depth 11 \
+  --trees 2000 \
+  --classes 3 \
+  --features 500) )
```
[the remark this answers is not in the document]
Comment (wphicks) [reply qa/test_entrypoint.sh:12]: Yep, I agree entirely! To be honest, I forgot that I'd left `run_ci_local.sh` in, but it's a good thing to have, so let me break that out appropriately.

File: qa/L0_e2e/test_model.py (hunk 1 above)
Comment (wphicks) [reply qa/L0_e2e/test_model.py:222]: This is actually a deliberate choice, though one that deserves a comment. By sandwiching the batch size `1` runs in between the larger runs, it increases the likelihood that the server will have to correctly handle inputs of different sizes in the same (server-side) batch of requests, which seems like an important thing to test. Let me add a comment explaining that.

Comment (hcho3) [reply qa/L0_e2e/test_model.py]: Thanks! I was wondering about why 1 was inserted in the middle of the list.

File: qa/test_entrypoint.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply qa/test_entrypoint.sh:12]: Great, I'll take another look when you update the local script.

[the remark this answers is not in the document]
Comment (wphicks) [reply qa/test_entrypoint.sh]: Just updated to combine these into a single script with the option to run locally with `LOCAL=1`. Thanks for catching that!

File: qa/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,41 @@
+ARG BASE_IMAGE=gitlab-master.nvidia.com:5005/whicks/triton_fil_backend
```
Comment (deadeyegoodwin) [inline qa/Dockerfile:1]: Shouldn't have any gitlab-master paths in here.

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**